Backend/src/controllers/auth.controller.ts

```diff
@@ -0,0 +1,102 @@
+};
+
+const register = async (req: Request, res: Response) => {
+  const { email, password, fullName, role } = req.body;
+  if (!email || !fullName || !password) {
+    return res.status(422).json({ message: 'The fields email, fullNamea and password are required' });
+  }
+
+  try {
+    const user = await User.findOne({ email });
+    if (user) {
+      return res.status(400).send('User already exist');
+    }
+
+    const hashedPassword = await bcrypt.hash(password, 10);
+    const newUser: UserInput = {
+      email,
+      password: hashedPassword,
+      enabled: true,
+      fullName,
+      role: role,
+    };
+    const userCreated = await User.create(newUser);
+    res.status(201).json({ name: userCreated.fullName, email: userCreated.email, _id: userCreated.id });
+  } catch (err) {
+    res.status(500).send('Server error');
+  }
+};
+
+const login = async (req: Request, res: Response) => {
+  const { email, password } = req.body;
+  try {
+    const user = await User.findOne({ email });
+    if (!user) {
+      return res.status(400).send('Invalid Email credentials');
+    }
+    const isMatch = await bcrypt.compare(password, user.password);
+    if (!isMatch) {
+      return res.status(400).send('Invalid Password credentials');
+    }
+
+    // delete other tokens that belogs to this user
+    await Token.deleteMany({ userId: user._id });
+
+    const accessToken = generateToken(user.id, 'access_token_secret', '1d');
+    const refreshToken = generateToken(user.id, 'refresh_token_secret', '7d');
+
+    const tokenDoc = new Token({
+      userId: user._id,
+      accessToken: accessToken,
+      refreshToken: refreshToken,
+      expiryDate: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
+    });
+    await tokenDoc.save();
+
+    res.json({ accessToken, refreshToken });
+  } catch (err) {
+    res.status(500).send('Server error');
+  }
+};
+
+const refreshToken = async (req: Request, res: Response) => {
+  const { token } = req.body;
+  if (!token) {
+    return res.status(401).send('Token is required');
+  }
+
+  try {
+    const tokenDoc = await Token.findOne({ refreshToken: token });
+    if (!tokenDoc) {
+      return res.status(401).send('Invalid token');
+    }
+
+    if (tokenDoc.expiryDate < new Date()) {
+      await Token.deleteOne({ token });
+      return res.status(401).send('Token expired');
+    }
+
+    const user = jwt.verify(token, 'refresh_token_secret');
+    const accessToken = generateToken((user as any).id, 'access_token_secret', '15m');
+    const newRefreshToken = generateToken((user as any).id, 'refresh_token_secret', '7d');
+
+    tokenDoc.accessToken = accessToken;
+    tokenDoc.refreshToken = newRefreshToken;
+    tokenDoc.expiryDate = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);
+    await tokenDoc.save();
+
+    res.json({ accessToken, refreshToken: newRefreshToken });
+  } catch (err) {
+    res.status(500).send('Server error');
+  }
+};
+
+export { login, register, refreshToken };
```

Comment:
Keep empty new line after the last line as a best practice. Do this in all the files.

Reply:
okay, got it

Reply:
Following files still do not have empty new line,

Backend/src/constants/constants.ts

Backend/src/constants/frontend.server.constants.ts

Backend/src/controllers/dto/error.dto.ts

Backend/src/middleware/auth.middleware.ts

Backend/src/middleware/cluster.middleware.ts

Backend/src/models/user.model.ts

Backend/src/routes/auth.routes.ts

Backend/src/routes/cluster.routes.ts

Backend/src/routes/graph.routes.ts

Backend/src/routes/user.routes.ts

Backend/src/databaseConnection.ts

Backend/.env.example

Frontend/src/app/auth/page.tsx

Frontend/src/app/clusters/[id]/access-management/page.tsx

Frontend/src/app/clusters/[id]/access-management/page.tsx

Frontend/src/app/clusters/[id]/instance/page.tsx

Frontend/src/app/clusters/[id]/page.tsx

Frontend/src/app/setup/page.tsx

Frontend/src/app/not-found.tsx

Frontend/src/components/auth/login-form.tsx

Frontend/src/components/cluster-details/cluster-registration-form.tsx

Frontend/src/components/cluster-details/user-registration-form.tsx

Frontend/src/components/graph-panel/hadoop-upload-modal.tsx

Frontend/src/components/graph-panel/kafka-upload-modal.tsx

Frontend/src/components/guides/about-jasmine.tsx

Frontend/src/components/setup/admin-profile.tsx

Frontend/src/components/setup/cluster-profile.tsx

Frontend/src/components/setup/last-step.tsx

Frontend/src/components/setup/welcome-screen.tsx

Frontend/src/data/side-menu-data.tsx

Frontend/src/data/user-data.tsx

Frontend/src/hooks/useAccessToken.tsx

Frontend/src/layouts/header/header.tsx

Frontend/src/redux/features/clusterData.ts

Frontend/src/redux/hook.ts

Frontend/src/routes/page-routes.ts

Frontend/src/services/analyzer-service.tsx

Frontend/src/services/axios.tsx

Frontend/src/services/cluster-service.tsx

Frontend/src/services/graph-service.tsx

Frontend/src/services/user-service.tsx

Frontend/src/types/user-types.ts

Frontend/src/utils/user-provider.tsx

Frontend/Dockerfile

docker-compose.yml

Reply:
empty new line added for above files

Reply:
Ok